enh: new job creation model

carlosgrillet opened the pull request:
<!--
  Before submitting a Pull Request, please ensure you've done the following:
  - 👷‍♀️ Create small PRs whenever possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive PRs and commit messages*.
  - 📗 Update any related documentation.

  * - We recommend the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format for commit messages and PR titles as a best practice.
-->

## Summary
<!--
  Example:
  This [adds|removes|updates|fixes|replaces] the [feature|bugfix|documentation|etc] ...
-->

For the packagebundle, when a new resource of this type is added, a new Job will spawn in the cluster and will:
- `initContainer`: download the package
- `container`: build the package

The job is now created with a path to store the repo also some improvements to the job have been made

- added a security context to improve security (non-root)
- resource requests and limits added
- task divided into two containers:
  - init container will download the package on the repository
  - container will run the necessary commands to build the package
- the user can now set the image for downloading and the image for building
  the packages. Implemented as `packagebundle.spec.config`

### Added/Updated tests?

- [X] Yes
- [ ] No, and this is why: _Please replace this line with details on why tests have not been included._

## PR info
<!--
  Example:
  PR link: #n <-number of PR
-->

PR link: #

carlosgrillet committed:
enh: new job creation model

The job is now created with a path to store the repo
also some improvements to the job have been made

- added a security context to improve security (non-root)
- resource requests and limits added
- task divided into two containers:
  - init container will download the package on the repository
  - container will run the necessary commands to build the package

carlosgrillet committed:
add: new packagebundle phase

- Loaded phase indicates that the packages have been downloaded
  and successfuly loaded int othe NSO targetName

carlosgrillet committed:
change: nso image, nso replicas, pb url and path

- NSO image changed to a privated one for development
- NSO replicas set to 1:
    This change is important for the following reason.
    When you multiple NSO instances cannot share the same
    PVC with the packages, unless your k8s cluster default
    storageClass supports RWX access mode.
- New real package used for tesing on packagebundle URL
- Path changed to the corresponding pakcages path on the new package

carlosgrillet committed:
add: new controller image and permissions

- the controller's docker image was set to the public one on
  my docker hub account
- controller image tag changed to 'alpha' while still on
  early development
- new permissions granted for packagebundle controller:
  - packagebundle controller can now get, list, patch, update and watch
    nsoes

carlosgrillet committed:
add: new normalizer function for path strings

carlosgrillet committed:
feat: mount packages on NSO instance

carlosgrillet committed:
feat: job updated for downloading and building pakcages

- now jobs use two containers. an initcontainer for downloading the
  packages, and a container with a NSO build image to compile all the
  packages under the folder `path` specified in the packagebundle
- security context removed from the build container
- new command for compiling all the packages on builder container

carlosgrillet committed:
test: updated test for new features

carlosgrillet committed:
update: packages

carlosgrillet committed:
update: golang builder image updated to version 1.25

carlosgrillet committed:
add: new packagebundle status field

- loading is added as a new state to indicate that the packages are
  being loaded into the NSO instance

carlosgrillet committed:
update: EnvVar field to version corev1 v0.35.0

- as part of the upgrade of the libs. k8s.io/api was upgraded to version
  0.35.0. This changes introduces new fields for EnvVars

carlosgrillet committed:
feat: new Config field added to NSOReconciler struct

- Initialized from the Manager, this is the Kubernetes
  REST client configuration containing cluster connection details
  (API server URL, authentication credentials, etc.). This is used
  to create a PodExecutor

carlosgrillet committed:
add: new permissions added for execute commands in pods

carlosgrillet committed:
feat!: new NSO package management system added.

Add end-to-end package lifecycle management:
- Download packages via git in init container, build with NSO compiler in main container
- Automatically mount built packages to target NSO instance via PVC
- Execute package reload command in NSO pod using Kubernetes exec API
- Track status through phases: downloading → building → mounting → loading

Technical changes:
- Add PodExecutor for remote command execution in pods
- Implement two-stage job: git clone + make build
- Add security contexts and resource limits to jobs
- Introduce sentinel errors for pod restart coordination
- Add helper functions for consistent resource naming

carlosgrillet committed:
test: new tests added for resources

carlosgrillet committed:
ci: update golangci-lint to version v2.7.2

cassandrabarreto commented on internal/controller/resources.go:
```diff
@@ -226,17 +250,28 @@ func (r *PackageBundleReconciler) newJob(ctx context.Context, pb *nsov1alpha1.Pa
-							fmt.Sprintf("cd /packages && git clone %s", pb.Spec.Source.Url),
-						},
+						Name:       "builder",
+						Image:      "carlosgrillet/cisco-nso:6.1.19-build",
```
The image name should be taken from the sample. It should not be a hardcoded value.

cassandrabarreto commented on internal/controller/resources_test.go:
```diff
@@ -413,24 +413,44 @@ var _ = Describe("Resource Creation Functions", func() {
-
-				// Check volume mounts
+				Expect(container.Name).To(Equal("builder"))
+				Expect(container.Image).To(Equal("carlosgrillet/cisco-nso:6.1.19-build"))
```
It should not be hardcoded.

cassandrabarreto commented on config/samples/orchestration.cisco.com_v1alpha1_nso.yaml:
```diff
@@ -6,9 +6,9 @@ metadata:
   name: nso-sample
 spec:
-  image: containers.cisco.com/cisco-nso/cisco-nso-prod:6.1.11
+  image: carlosgrillet/cisco-nso:6.1.19-prod
```
This should not be hardcoded. Maybe we can keep an image from containers.cisco that corresponds to NSO 6.1.19.

cassandrabarreto commented on internal/controller/packagebundle_controller_test.go:
```diff
@@ -118,7 +118,8 @@ var _ = Describe("PackageBundle Controller", func() {
 			Expect(err).NotTo(HaveOccurred())
-			Expect(job.Spec.Template.Spec.Containers[0].Image).To(Equal("alpine/git"))
+			Expect(job.Spec.Template.Spec.InitContainers[0].Image).To(Equal("alpine/git"))
+			Expect(job.Spec.Template.Spec.Containers[0].Image).To(Equal("carlosgrillet/cisco-nso:6.1.19-build"))
```
Tests should not include a hardcoded value for the image name.

cassandrabarreto reviewed (changes requested):
I tested the functionality. It works. However, we must get rid of hardcoded values in code.

carlosgrillet committed:
enh: packagebundle download and build image

- packagebundle download and build container now have dynamic names.
  - download container image has the name
    "<packagebundlename>-downloader"
  - build container image has the name
    "<packagebundlename>-build"
- image for packagebundle download and build containers
  can now be set by user via the manifest.

carlosgrillet committed:
update: new config field added with download and build configs

carlosgrillet committed:
enh: packagebundle config field

- two new types added to packagebundle types.
  - PackageConfig: holds the configuration for download an build
    containers. These fileds are of type ContainerParams.
  - ContainerParams: Here is where the config for the container is
    placed. At this time, only 'image' is required, but here we can
    add more fields like "args", or "command" if needed.

carlosgrillet committed:
fix: removed orphan comment from packagebundle types

carlosgrillet committed:
test: test cases updated for packagebuncle controller and resources

carlosgrillet committed:
update: manifest regenerated. Kubebuilder auto-generated

carlosgrillet committed:
fix: added retry logic to resolve resource update conflicts

Resource update operations in controllers were failing with "object has been
modified" errors due to concurrent reconciliation loops. This fix implements
retry logic with conflict resolution for all update operations.

Changes:
- updatePackageBundlePhase: now checks all status fields (phase, message,
  jobName) before updating to avoid unnecessary updates that cause conflicts
- ensureObjectExists: wrapped Update call with RetryOnConflict to handle
  StatefulSet/Service modifications during reconciliation
- mountPVCToNSO: added retry logic when updating NSO spec with volume mounts,
  includes idempotent check to prevent duplicate volume additions
- Moved error logging outside retry loops to reduce noise

All update operations now fetch the latest resource version on each retry
attempt, preventing conflicts when resources are modified between read and
update operations.

cassandrabarreto reviewed (approved):
Tested it on a local environment. No issues detected. Awesome work. :)
